src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -71,64 +70,55 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
-                // passed to a nested construction call. This needs to affect the current call to ensure side effects are
-                // preserved, but shouldn't affect any subsequent calls.
-                BoundExpression? appendReceiver = builderTemp;
-                Debug.Assert(appendCall.Method.RequiresInstanceReceiver);
-                var argRefKindsOpt = appendCall.ArgumentRefKindsOpt;
-
-                var rewrittenArguments = VisitArguments(
-                    appendCall.Arguments,
-                    appendCall.Method,
-                    appendCall.ArgsToParamsOpt,
-                    argRefKindsOpt,
-                    ref appendReceiver,
-                    out ArrayBuilder<LocalSymbol>? temps);
-
-                var rewrittenAppendCall = MakeArgumentsAndCall(
-                    appendCall.Syntax,
-                    appendReceiver,
-                    appendCall.Method,
-                    rewrittenArguments,
-                    appendCall.ArgumentRefKindsOpt,
-                    appendCall.Expanded,
-                    appendCall.InvokedAsExtensionMethod,
-                    appendCall.ArgsToParamsOpt,
-                    appendCall.ResultKind,
-                    appendCall.Type,
-                    temps,
-                    appendCall);
-
-                Debug.Assert(usesBoolReturn == (rewrittenAppendCall.Type!.SpecialType == SpecialType.System_Boolean));
-                resultExpressions.Add(rewrittenAppendCall);
+                if (part is BoundCall call)
+                {
+                    Debug.Assert(call.Type.SpecialType == SpecialType.System_Boolean == usesBoolReturns);
+                    resultExpressions.Add((BoundExpression)VisitCall(call));
+                }
+                else if (part is BoundDynamicInvocation dynamicInvocation)
+                {
+                    resultExpressions.Add(VisitDynamicInvocation(dynamicInvocation, resultDiscarded: !usesBoolReturns));
+                }
+                else
+                {
+                    throw ExceptionUtilities.UnexpectedValue(part.Kind);
+                }
             }
 
+            RemovePlaceholderReplacement(data.ReceiverPlaceholder);
+
             if (appendShouldProceedLocal is not null)
             {
                 RemovePlaceholderReplacement(data.ArgumentPlaceholders[^1]);
             }
 
-            if (usesBoolReturn)
+            if (usesBoolReturns)
             {
                 // We assume non-bool returns if there was no parts to the string, and code below is predicated on that.
                 Debug.Assert(!node.Parts.IsEmpty);
                 // Start the sequence with appendProceedLocal, if appropriate
                 BoundExpression? currentExpression = appendShouldProceedLocal;
 
+                var boolType = _compilation.GetSpecialType(SpecialType.System_Boolean);
```

Comment:
nit (ie. not blocking): we'll want a test where we make the type missing

Reply:
Of note, we required this type during initial binding.